File: ios/TNMInterceptorManager.swift
//
//  TNMInterceptorManager.swift
//  TiNetworkManager
//
//  Created by Douglas Alves on 02/01/26.
//


/**
 * Ti.Network.Manager - Interceptor Manager
 * Handles request/response interceptors
 * Allows global middleware for auth, logging, error handling, etc.
 */

import TitaniumKit
import Foundation

class TNMInterceptorManager {
    
    // MARK: - Properties
    
    private var requestInterceptors: [KrollCallback] = []
    private var responseInterceptors: [KrollCallback] = []
    
    var requestInterceptorCount: Int {
        return requestInterceptors.count
    }
    
    var responseInterceptorCount: Int {
        return responseInterceptors.count
    }
    
    // MARK: - Public Methods
    
    /**
     * Add request interceptor
     */
    func addRequestInterceptor(_ callback: KrollCallback) {
        requestInterceptors.append(callback)
    }
    
    /**
     * Add response interceptor
     */
    func addResponseInterceptor(_ callback: KrollCallback) {
        responseInterceptors.append(callback)
    }
    
    /**
     * Remove all request interceptors
     */
    func clearRequestInterceptors() {
        requestInterceptors.removeAll()
        
        TNMLogger.info("All request interceptors cleared", feature: "Interceptor")
    }
    
    /**
     * Remove all response interceptors
     */
    func clearResponseInterceptors() {
        responseInterceptors.removeAll()
        
        TNMLogger.info("All response interceptors cleared", feature: "Interceptor")
    }
    
    /**
     * Intercept outgoing request
     */
    func interceptRequest(
        url: String,
        method: String,
        headers: inout [String: String],
        body: inout String?
    ) {
        guard !requestInterceptors.isEmpty else { return }
        
        TNMLogger.Interceptor.requestIntercepted(url: url, method: method)
        
        for interceptor in requestInterceptors {
            // Create config object
            var config: [String: Any] = [
                "url": url,
                "method": method,
                "headers": headers
            ]
            
            if let body = body {
                config["body"] = body
            }
            
            // Call interceptor
            if let result = interceptor.call([config], thisObject: nil) as? [String: Any] {
                // Update headers if modified
                if let modifiedHeaders = result["headers"] as? [String: String] {
                    headers = modifiedHeaders
                }
                
                // Update body if modified
                if let modifiedBody = result["body"] as? String {
                    body = modifiedBody
                } else if result["body"] is NSNull {
                    body = nil
                }
            }
        }
    }
    
    /**
     * Intercept incoming response
     */
    func interceptResponse(
        statusCode: inout Int,
        headers: inout [String: String],
        body: String?
    ) {
        guard !responseInterceptors.isEmpty else { return }
        
        TNMLogger.Interceptor.responseIntercepted(statusCode: statusCode)
        
        for interceptor in responseInterceptors {
            // Create response object
            var response: [String: Any] = [
                "status": statusCode,
                "statusCode": statusCode,
                "headers": headers
            ]
            
            if let body = body {
                response["body"] = body
            }
            
            // Call interceptor
            if let result = interceptor.call([response], thisObject: nil) as? [String: Any] {
                // Update status code if modified
                if let modifiedStatus = result["status"] as? Int {
                    statusCode = modifiedStatus
                } else if let modifiedStatus = result["statusCode"] as? Int {
                    statusCode = modifiedStatus
                }
                
                // Update headers if modified
                if let modifiedHeaders = result["headers"] as? [String: String] {
                    headers = modifiedHeaders
                }
            }
        }
    }
}

File: android/src/ti/network/manager/TNMInterceptorManager.java
/**
 * Ti.Network.Manager - Interceptor Manager
 * Manages request and response interceptors
 * Uses OkHttp's Interceptor interface
 */

package ti.network.manager;

import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;
import org.appcelerator.kroll.KrollDict;
import org.appcelerator.kroll.KrollFunction;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class TNMInterceptorManager {

    private List<KrollFunction> requestInterceptors = new ArrayList<>();
    private List<KrollFunction> responseInterceptors = new ArrayList<>();

    public TNMInterceptorManager() {
        TNMLogger.info("Interceptor manager initialized", "Interceptor");
    }

    /**
     * Add request interceptor
     */
    public void addRequestInterceptor(KrollFunction callback) {
        requestInterceptors.add(callback);
        TNMLogger.Interceptor.requestInterceptorAdded(requestInterceptors.size());
    }

    /**
     * Add response interceptor
     */
    public void addResponseInterceptor(KrollFunction callback) {
        responseInterceptors.add(callback);
        TNMLogger.Interceptor.responseInterceptorAdded(responseInterceptors.size());
    }

    /**
     * Create OkHttp request interceptor
     */
    public Interceptor createRequestInterceptor() {
        return new Interceptor() {
            @Override
            public Response intercept(Chain chain) throws IOException {
                Request request = chain.request();

                // Call all request interceptors
                for (KrollFunction callback : requestInterceptors) {
                    KrollDict config = new KrollDict();
                    config.put("url", request.url().toString());
                    config.put("method", request.method());

                    // Convert headers
                    Map<String, String> headers = new HashMap<>();
                    for (String name : request.headers().names()) {
                        headers.put(name, request.header(name));
                    }
                    config.put("headers", headers);

                    // Call interceptor
                    Object result = callback.call(null, new Object[] { config });

                    if (result instanceof KrollDict) {
                        KrollDict modifiedConfig = (KrollDict) result;

                        // Rebuild request with modified config
                        Request.Builder builder = request.newBuilder();

                        if (modifiedConfig.containsKey("headers")) {
                            @SuppressWarnings("unchecked")
                            Map<String, String> modifiedHeaders =
                                    (Map<String, String>) modifiedConfig.get("headers");

                            for (Map.Entry<String, String> entry : modifiedHeaders.entrySet()) {
                                builder.header(entry.getKey(), entry.getValue());
                            }
                        }

                        request = builder.build();
                    }
                }

                return chain.proceed(request);
            }
        };
    }

    /**
     * Create OkHttp response interceptor
     */
    public Interceptor createResponseInterceptor() {
        return new Interceptor() {
            @Override
            public Response intercept(Chain chain) throws IOException {
                Response response = chain.proceed(chain.request());

                // Call all response interceptors
                for (KrollFunction callback : responseInterceptors) {
                    KrollDict responseDict = new KrollDict();
                    responseDict.put("statusCode", response.code());

                    // Convert headers
                    Map<String, String> headers = new HashMap<>();
                    for (String name : response.headers().names()) {
                        headers.put(name, response.header(name));
                    }
                    responseDict.put("headers", headers);

                    // Call interceptor
                    callback.call(null, new Object[] { responseDict });
                }

                return response;
            }
        };
    }

    /**
     * Clear all interceptors
     */
    public void clearAllInterceptors() {
        requestInterceptors.clear();
        responseInterceptors.clear();
        TNMLogger.info("All interceptors cleared", "Interceptor");
    }
}